File: LeetCode/763. Partition Labels.js
// https://leetcode.com/problems/partition-labels/description/

const { test } = require('../test');

const partitionLabels = (str) => {
  const lastCharIndex = new Map();

  const result = [];

  for (let i = 0; i < str.length; i++) {
    lastCharIndex.set(str[i], i);
  }

  let largestIndex = Number.NEGATIVE_INFINITY;

  let part = '';

  for (let i = 0; i < str.length; i++) {
    largestIndex = Math.max(largestIndex, lastCharIndex.get(str[i]));

    part += str[i];

    if (largestIndex === i) {
      result.push(part.length);

      part = '';
      largestIndex = Number.NEGATIVE_INFINITY;
    }
  }

  return result;
};

test(partitionLabels, [
  {
    input: ['ababcbacadefegdehijhklij'],
    expected: [9, 7, 8].join('-'),
  },
  {
    input: ['eccbbbbdec'],
    expected: [10].join('-'),
  },
]);

File: LeetCode/763. Partition Labels.py
# https://leetcode.com/problems/partition-labels/description/

from test import test


class Solution:
    def partitionLabels(self, string: str) -> list[int]:
        last_char_index: dict[str, int] = {}

        for index, char in enumerate(string):
            last_char_index[char] = index

        result: list[int] = []

        last_largest_index = -1
        part = ""

        for index, char in enumerate(string):
            part += char

            last_largest_index = max(last_largest_index, last_char_index[char])

            if last_largest_index == index:
                result.append(len(part))

                part = ""

        return result


solution = Solution()


test(
    solution.partitionLabels,
    [
        {
            "input": ["ababcbacadefegdehijhklij"],
            "expected": [9, 7, 8],
        },
        {
            "input": ["eccbbbbdec"],
            "expected": [10],
        },
    ],
)
